File: apps/backend/src/rhesis/backend/app/schemas/test_configuration.py
from typing import Any, Dict, List, Optional

from pydantic import UUID4, BaseModel, Field

from rhesis.backend.app.schemas import Base
from rhesis.backend.app.schemas.tag import TagRead


# TestConfiguration schemas
class TestConfigurationBase(Base):
    endpoint_id: UUID4
    category_id: Optional[UUID4] = None
    topic_id: Optional[UUID4] = None
    prompt_id: Optional[UUID4] = None
    use_case_id: Optional[UUID4] = None
    test_set_id: Optional[UUID4] = None
    user_id: Optional[UUID4] = None
    organization_id: Optional[UUID4] = None
    status_id: Optional[UUID4] = None
    attributes: Optional[dict] = None


class TestConfigurationCreate(TestConfigurationBase):
    pass


class TestConfigurationUpdate(TestConfigurationBase):
    endpoint_id: Optional[UUID4] = None


class TestConfiguration(TestConfigurationBase):
    pass


class TestSetReference(Base):
    id: UUID4
    name: Optional[str] = None
    description: Optional[str] = None
    counts: Optional[Dict[str, Any]] = None
    user_id: Optional[UUID4] = None
    organization_id: Optional[UUID4] = None
    status_id: Optional[UUID4] = None
    attributes: Optional[Dict[str, Any]] = None
    tags: Optional[List[TagRead]] = None


class EndpointReference(Base):
    id: UUID4
    name: Optional[str] = None
    description: Optional[str] = None
    user_id: Optional[UUID4] = None
    organization_id: Optional[UUID4] = None
    status_id: Optional[UUID4] = None
    tags: Optional[List[TagRead]] = None


# The detailed model with expanded relations
class TestConfigurationDetail(TestConfiguration):
    # The factory-generated schema overrides endpoint_id to Optional (the
    # base schema declares it required) -- preserved here for parity.
    id: UUID4
    endpoint_id: Optional[UUID4] = None
    test_set: Optional[TestSetReference] = None
    endpoint: Optional[EndpointReference] = None


class TestConfigurationExecutionRequest(BaseModel):
    """Request model for test configuration execution."""

    experiment_id: Optional[UUID4] = Field(
        default=None,
        description=(
            "When set (optionally with version / environment), intent is stored on the "
            "test configuration and resolved into a run snapshot at queue time."
        ),
    )
    version: Optional[str] = None
    environment: Optional[str] = None

File: apps/frontend/src/utils/api-client/interfaces/test-configuration.ts
import { UUID } from 'crypto';
import { TestSet } from './test-set';
import { Endpoint } from './endpoint';

/**
 * Enum for tracking the source of metrics used in a test execution.
 * Must match the backend MetricsSource enum in schemas/test_set.py
 */
export enum MetricsSource {
  REQUIREMENT = 'requirement',
  TEST_SET = 'test_set',
  EXECUTION_TIME = 'execution_time',
}

/**
 * Helper to get display label for MetricsSource
 */
export function getMetricsSourceLabel(source: MetricsSource | string): string {
  switch (source) {
    case MetricsSource.REQUIREMENT:
      return 'Requirement';
    case MetricsSource.TEST_SET:
      return 'Test Set';
    case MetricsSource.EXECUTION_TIME:
      return 'Execution-Time';
    default:
      return 'Unknown';
  }
}

/**
 * Execution metric stored in test configuration attributes
 */
export interface ExecutionMetric {
  id: string;
  name: string;
  scope?: string[];
}

/**
 * Typed test configuration attributes
 */
export interface TestConfigurationAttributes {
  execution_mode?: 'Parallel' | 'Sequential';
  metrics?: ExecutionMetric[];
  metrics_source?: MetricsSource;
  [key: string]: unknown;
}

// Test Configuration interfaces
export interface TestConfigurationBase {
  endpoint_id: UUID;
  test_set_id: UUID;
  user_id: UUID;
  attributes?: TestConfigurationAttributes;
}

export interface TestConfigurationCreate extends Omit<
  TestConfigurationBase,
  'user_id'
> {
  user_id?: UUID;
}

export type TestConfigurationUpdate = Partial<TestConfigurationBase>;

export interface TestConfiguration extends TestConfigurationBase {
  id: UUID;
}

export interface TestConfigurationDetail extends TestConfiguration {
  endpoint?: Endpoint;
  test_set?: TestSet;
}
